jobs/integration/validation.py

```diff
@@ -2306,68 +2372,61 @@ async def test_nagios(model, nagios):
-                log("Found critical errors:")
-                for s in found_master + found_worker:
-                    log(" - {}".format(s))
-                break
-        await asyncio.sleep(30)
+    try:
+        # 1) break all the things
+        log("breaking api server")
+        await control_plane.set_config({"api-extra-args": "broken=true"})
 
-    # 6) fix api and wait for settle
-    log("Fixing API server")
-    await masters.set_config({"api-extra-args": ""})
-    await nagios.wait_for_alerts(critical=lambda c: not c)
+        # 2) make sure nagios is complaining for kubernetes-control-plane
+        #    AND kubernetes-worker
+        log("Verifying complaints")
+        await retry_async_with_timeout(
+            nagios.critical_alerts_by_app,
+            (
+                "kubernetes-control-plane",
+                "kubernetes-worker",
+            ),
+            timeout_insec=10 * 60,
+            timeout_msg="Failed to find critical errors in control-plane and worker after 10m.",
+            retry_interval_insec=30,
+        )
+    finally:
+        # 3) fix api
+        log("Fixing API server")
+        await control_plane.set_config({"api-extra-args": ""})
 
-    # 7) break worker
-    log("Breaking workers")
-    workers = masters = model.applications["kubernetes-worker"]
-    await workers.set_config({"kubelet-extra-args": "broken=true"})
+    # wait for complaints to go away
+    await nagios.wait_for_settle(
+        stage="restoring API Server",
+        timeout_insec=10 * 60,
+    )
 
-    # 8) verify nagios is complaining about worker
-    log("Verifying complaints")
-    while True:
-        alerts = await nagios.wait_for_alerts(critical=lambda c: len(c) > 0)
-        criticals = alerts["critical"]
+    try:
+        # 4) break worker
+        log("Breaking workers")
+        workers = model.applications["kubernetes-worker"]
+        await workers.set_config({"kubelet-extra-args": "broken=true"})
 
-        if criticals:
-            found_worker = []
-            for c in criticals:
-                for link in c.find_all("a", recursive=False):
-                    if "kubernetes-worker" in link.string:
-                        found_worker.append(link.string)
-                        break
-            if found_worker:
-                log("Found critical errors:")
-                for s in found_worker:
-                    log(" - {}".format(s))
-                break
-        await asyncio.sleep(30)
+        # 5) verify nagios is complaining about worker
+        log("Verifying complaints")
+        await retry_async_with_timeout(
+            nagios.critical_alerts_by_app,
+            ("kubernetes-worker",),
+            timeout_insec=10 * 60,
+            timeout_msg="Failed to find critical errors in worker after 10m.",
+            retry_interval_insec=30,
+        )
+    finally:
+        # 9) Fix worker
+        await workers.set_config({"kubelet-extra-args": ""})
```

Comment:
similar to the first test, we're gonna break the kubelets on all the workers and look for critical failures on only the workers.  It's got 10 minutes to fail -- GO!